File: samples/dotnet-isolated/eventhub/KafkaTriggerManyWithHeaders.cs
using System;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Extensions.Logging;
using Microsoft.Azure.Functions.Worker.Http;
using Newtonsoft.Json.Linq;


namespace Eventhub
{
    public class KafkaTriggerManyWithHeaders
    {
        [Function("KafkaTriggerManyWithHeaders")]
        public static void Run(
            [KafkaTrigger("BrokerList",
                          "topic",
                          Username = "$ConnectionString",
                          Password = "EventHubConnectionString",
                          Protocol = BrokerProtocol.SaslSsl,
                          AuthenticationMode = BrokerAuthenticationMode.Plain,
                          ConsumerGroup = "$Default",
                          IsBatched = true)] string[] events, FunctionContext context)
        {
            foreach (var kevent in events)
            {
                var eventJsonObject = JObject.Parse(kevent);
                var logger = context.GetLogger("KafkaFunction");
                logger.LogInformation($"C# Kafka trigger function processed a message: {eventJsonObject["Value"]}");

                var headersJArr = eventJsonObject["Headers"] as JArray;
                logger.LogInformation("Headers for this event: ");
                foreach (JObject header in headersJArr)
                {
                    logger.LogInformation($"{header["Key"]} {System.Text.Encoding.UTF8.GetString((byte[])header["Value"])}");

                }
            }
        }
    }
}

File: samples/javascript-v4/src/functions/kafkaTriggerManyWithHeaders.js
const { app } = require("@azure/functions");

async function kafkaTriggerManyWithHeaders(events, context) {
  for (const event of events) {
    context.log("Event Offset: " + event.Offset);
    context.log("Event Partition: " + event.Partition);
    context.log("Event Topic: " + event.Topic);
    context.log("Event Timestamp: " + event.Timestamp);
    context.log("Event Value (as string): " + event.Value);

    let event_obj = JSON.parse(event.Value);

    context.log("Event Value Object: ");
    context.log("   Value.registertime: ", event_obj.registertime.toString());
    context.log("   Value.userid: ", event_obj.userid);
    context.log("   Value.regionid: ", event_obj.regionid);
    context.log("   Value.gender: ", event_obj.gender);

    // Headers
    for (let index in event.Headers) {
      context.log(
        `Header ${index} - index: ${event.Headers[index].Key}, value: ${event.Headers[index].Value}`
      );
    }
  }
}

// eventhub
app.generic("kafkaTriggerManyWithHeaders", {
  trigger: {
    type: "kafkaTrigger",
    direction: "in",
    name: "events",
    topic: "topic",
    brokerList: "%BrokerList%",
    username: "$ConnectionString",
    password: "EventHubConnectionString",
    consumerGroup: "$Default",
    protocol: "saslSsl",
    authenticationMode: "plain",
    dataType: "string",
    cardinality: "MANY",
  },
  handler: kafkaTriggerManyWithHeaders,
});

// confluent
// app.generic("Kafkatrigger", {
//   trigger: {
//     type: "kafkaTrigger",
//     direction: "in",
//     name: "event",
//     topic: "topic",
//     brokerList: "%BrokerList%",
//     username: "%ConfluentCloudUserName%",
//     password: "%ConfluentCloudPassword%",
//     consumerGroup: "$Default",
//     protocol: "saslSsl",
//     authenticationMode: "plain",
//     dataType: "string",
//     cardinality: "MANY"
//   },
//   handler: kafkaTriggerManyWithHeaders,
// });
